dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,11 +612,671 @@ class FunctionsJsonArray : public IFunction
+        auto & data_to = col_to->getChars();
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
+
+        // In raw function test, input_tidb_tp/output_tidb_tp is nullptr.
+        if (collator && collator->isBinary())
+        {
+            if (unlikely(input_tidb_tp == nullptr))
+            {
+                doExecuteForBinary<false>(data_to, offsets_to, input_source, input_tidb_tp->tp(), -1, block.rows());
+            }
+            else if (input_tidb_tp->tp() == TiDB::TypeString)
+            {
+                doExecuteForBinary<true>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+            else
+            {
+                doExecuteForBinary<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    input_tidb_tp->tp(),
+                    input_tidb_tp->flen(),
+                    block.rows());
+            }
+        }
+        else if ((unlikely(output_tidb_tp == nullptr)) || hasParseToJSONFlag(*output_tidb_tp))
+        {
+            if (from.column->isColumnNullable())
+            {
+                const auto & column_nullable = static_cast<const ColumnNullable &>(*from.column);
+                doExecuteForParsingJson<true>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    column_nullable.getNullMapData(),
+                    block.rows());
+            }
+            else
+            {
+                auto tmp_null_map = ColumnUInt8::create(0, 0);
+                doExecuteForParsingJson<false>(
+                    data_to,
+                    offsets_to,
+                    input_source,
+                    tmp_null_map->getData(),
+                    block.rows());
+            }
+        }
+        else
+        {
+            doExecuteForOthers(data_to, offsets_to, input_source, block.rows());
+        }
+
+        if (from.column->isColumnNullable())
+        {
+            auto null_map = static_cast<const ColumnNullable &>(*from.column).getNullMapColumnPtr();
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_to), std::move(null_map));
+        }
+        else
+        {
+            block.getByPosition(result).column = std::move(col_to);
+        }
+    }
+
+private:
+    template <bool is_binary_str>
+    static void doExecuteForBinary(
+        ColumnString::Chars_t & data_to,
+        ColumnString::Offsets & offsets_to,
+        const std::unique_ptr<IStringSource> & data_from,
+        UInt8 from_type_code,
+        Int32 flen,
+        size_t size)
+    {
+        size_t reserve_size = (size * (1 + 1)) + data_from->getSizeForReserve();
```

Comment:
Why need size * (1 + 1) here? I think size * 1 is enough?

Reply:
Because the reserve_size = `size  * (json_type + string end char 0 + value)`.